File: gto-support-api-retrofit2/src/main/java/org/ccci/gto/android/common/api/retrofit2/converter/JSONObjectConverterFactory.java
package org.ccci.gto.android.common.api.retrofit2.converter;

import static org.ccci.gto.android.common.retrofit2.converter.JSONObjectConverterFactory.INSTANCE;

import java.lang.annotation.Annotation;
import java.lang.reflect.Type;

import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import retrofit2.Converter;
import retrofit2.Retrofit;

/**
 * @deprecated Since v3.10.0, use {@link org.ccci.gto.android.common.retrofit2.converter.JSONObjectConverterFactory}
 * instead.
 */
@Deprecated
public final class JSONObjectConverterFactory extends Converter.Factory {
    @Override
    public Converter<?, RequestBody> requestBodyConverter(Type type, Annotation[] parameterAnnotations,
                                                          Annotation[] methodAnnotations, Retrofit retrofit) {
        return INSTANCE.requestBodyConverter(type, parameterAnnotations, methodAnnotations, retrofit);
    }

    @Override
    public Converter<ResponseBody, ?> responseBodyConverter(Type type, Annotation[] annotations,
                                                            Retrofit retrofit) {
        return INSTANCE.responseBodyConverter(type, annotations, retrofit);
    }
}

File: gto-support-retrofit2/src/main/kotlin/org/ccci/gto/android/common/retrofit2/converter/JSONObjectConverterFactory.kt
package org.ccci.gto.android.common.retrofit2.converter

import java.io.IOException
import java.lang.reflect.Type
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody
import okhttp3.ResponseBody
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import retrofit2.Converter
import retrofit2.Retrofit

object JSONObjectConverterFactory : Converter.Factory() {
    override fun requestBodyConverter(
        type: Type,
        parameterAnnotations: Array<Annotation>,
        methodAnnotations: Array<Annotation>,
        retrofit: Retrofit,
    ): Converter<*, RequestBody>? = when (type) {
        JSONObject::class.java, JSONArray::class.java -> JSONObjectRequestBodyConverter
        else -> null
    }

    override fun responseBodyConverter(type: Type, annotations: Array<Annotation>, retrofit: Retrofit) = when (type) {
        JSONObject::class.java -> JSONObjectResponseBodyConverter
        JSONArray::class.java -> JSONArrayResponseBodyConverter
        else -> null
    }

    private object JSONObjectRequestBodyConverter : Converter<Any?, RequestBody> {
        private val MEDIA_TYPE = "text/json; charset=UTF-8".toMediaType()

        override fun convert(value: Any?): RequestBody = RequestBody.create(MEDIA_TYPE, value?.toString().orEmpty())
    }

    private object JSONObjectResponseBodyConverter : Converter<ResponseBody, JSONObject> {
        override fun convert(value: ResponseBody) = try {
            JSONObject(value.string())
        } catch (e: JSONException) {
            throw IOException("Error parsing JSON", e)
        }
    }

    private object JSONArrayResponseBodyConverter : Converter<ResponseBody, JSONArray> {
        override fun convert(value: ResponseBody) = try {
            JSONArray(value.string())
        } catch (e: JSONException) {
            throw IOException("Error parsing JSON", e)
        }
    }
}
